src/include/dstore.h

```diff
@@ -182,4 +184,11 @@ int dstore_pwrite(struct dstore_obj *obj, off_t offset, size_t count,
  */
 int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
 		       size_t bs, char *buf);
+
+/** This API does the handling for truncating a file
+ *  @param[in] obj - An open object.
+ *  @param[in] vec - vector filled with offset and count.
+ *  @return 0 or -errno.
+ */
+int dstore_ptrunc(struct dstore_obj *obj, struct dstore_extent_vec *vec);
```

Comment:
What do you say about the following interface?

```c
enum dstore_obj_fallocate_mode {
  /* De-allocate storage space allocated by by Alloc or Write operations */
  DSTORE_OBJ_FALLOC_FREE,
};

/* Allocates or de-allocates synchronously the range specified by offset and len.
 * When the range is not aligned with the block size of the underlying storage,
 * it fills the leftovers with zeros.
 * TODO: Allocate mode is not implemented yet.
 * TODO: Unaligned ranges are not handled yet.
 * @param mode Operation to be perform (see enum ::dstore_obj_fallocate_mode).
 */
int dstore_fallocate(struct dstore_obj *obj, uint64_t mode, off_t offset, off_t len);
```

We don't need the ability to free a group of extents here: truncate() changes the file size continuously -- there are no gaps, so that it is enough to have only one extent.

As usually we have a choice here: either provide an interface that reassembles m0 client API (with operations, io/index vectors and callbacks) or an interface that reassembles simple POSIX-like API (single range, synchronous). Since the previous operations (pwrite, pread) are quite close to their POSIX counterparts, let's see if we can make our Alloc/Free to be similar to its POSIX equivalent (fallocate).

Reply:
I have done new posix formta implementation. I want to keep api separate for alloc and dealloc.